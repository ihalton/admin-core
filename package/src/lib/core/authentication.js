import { oidcLog } from "@axa-fr/react-oidc-context";

const makeAbsolute = (relativeUrl, rootUrl) => {
  const url = new URL(relativeUrl, rootUrl); // eslint-disable-line node/no-unsupported-features/node-builtins
  return url.href;
};

/**
 * Generate props for the Oidc AuthenticationProvider
 * @param {Object} config Config options
 * @param {String} config.oidcClientId OIDC client id
 * @param {String} config.oidcUrl OIDC url
 * @param {String} config.rootUrl Webapp root url
 * @returns {Object} An object containing props for the Oidc AuthenticationProvider
 */
export function getOidcProps({ oidcClientId, oidcUrl, rootUrl, ...otherProps }) {
  /* eslint-disable camelcase */
  const oidcConfiguration = {
    client_id: oidcClientId,
    redirect_uri: makeAbsolute("/authentication/callback", rootUrl),
    response_type: "code",
    post_logout_redirect_uri: rootUrl,
    scope: "openid",
    authority: oidcUrl,
    silent_redirect_uri: makeAbsolute("/authentication/silent_callback", rootUrl),
    automaticSilentRenew: true,
    loadUserInfo: true,
    triggerAuthFlow: true
  };
  /* eslint-enable camelcase */

  return {
    configuration: oidcConfiguration,
    // NONE, ERROR, WARN, INFO, DEBUG
    // Change to DEBUG temporarily if you're debugging an issue with login/logout/auth
    loggerLevel: oidcLog.NONE,
    // These are components for which the @axa-fr/react-oidc-context package shows
    // default text if we don't override these. We don't really need them since in
    // our situation they're only shown for a second.
    authenticating: () => null,
    callbackComponentOverride: () => null,
    notAuthenticated: () => null,
    notAuthorized: () => null,
    sessionLostComponent: () => null,
    ...otherProps
  };
}
