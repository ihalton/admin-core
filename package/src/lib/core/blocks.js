import Logger from "../../utils/logger";

export const BlocksTable = {}; // storage for separate elements of each block
export const BlockComponents = {};

/**
 * @example // Register a component and container(s) with a name.
 * // The raw component can then be extended or replaced.
 *
 * // Structure of a component in the list:
 *
 * BlocksTable.MyComponent = {
 *    name: 'MyComponent',
 *    hocs: [fn1, fn2],
 *    rawComponent: React.Component
 * }
 * @name registerComponent
 * @method
 * @memberof Components/Helpers
 * @param {Object} options The name of the component to register.
 * @param {Object} options.name The name of the component to register.
 * @param {Object} options.Component The name of the component to register.
 * @param {Object} options.region The name of the component to register.
 * @param {Object} options.priority The name of the component to register.
 * @param {Object} options.priority The name of the component to register.
 * @param {React.Component} rawComponent Interchangeable/extendable component.
 * @param {Function|Array} hocs The HOCs to wrap around the raw component.
 *
 * @returns {React.Component} returns the final wrapped component
 */
export function registerBlock(options) {
  const {
    name,
    Component,
    region,
    priority
  } = options;

  if (!region) {
    throw new Error("A region is required for registerBlock");
  }

  if (!name) {
    throw new Error("A name is required for registerBlock");
  }

  if (!Component) {
    throw new Error("A component is required for registerBlock");
  }

  if (!BlocksTable[region]) {
    BlocksTable[region] = {};
  }

  // store the component in the table
  BlocksTable[region][name] = {
    name,
    Component,
    region,
    priority
  };
}

/**
 * @name getBlock
 * @method
 * @summary Get a block registered with registerBlock({ name, component, hocs, region }).
 * @param {String} regionName The name of region the block belongs to.
 * @param {String} blockName The name of the block.
 * @returns {Function|React.Component} A (wrapped) React component
 * @memberof Components/Helpers
 */
export function getBlock(regionName, blockName) {
  const block = BlocksTable[regionName][blockName];

  if (!block) {
    throw new Error(`Block ${blockName} in region ${regionName} not registered.`);
  }

  return block.Component;
}

/**
 * @name getBlocks
 * @method
 * @summary Get a component registered with registerComponent(name, component, ...hocs).
 * @param {String} regionName The name of the region to get.
 * @returns {Function|React.Component} A (wrapped) React component
 * @memberof Components/Helpers
 */
export function getBlocks(regionName) {
  const region = BlocksTable[regionName];

  if (!region) {
    Logger.warn(`No blocks available for region named ${regionName}.`);
    return null;
  }

  const blocks = Object
    .values(region)
    .sort((blockA, blockB) => blockA.priority - blockB.priority)
    .map(({ Component }) => Component);

  return blocks;
}

/**
 * @name replaceBlock
 * @method
 * @summary Replace a Reaction component with a new component and optionally add one or more higher order components.
 * This function keeps track of the previous HOCs and wraps the new HOCs around previous ones
 * @param {Object} options Object containing block information
 * @param {String} options.region The region of the block that will be replaced
 * @param {String} options.block The name of the block that will be replaced
 * @param {React.Component} options.component Interchangeable/extendable component.
 * @returns {Function|React.Component} A component callable with Components[name]
 * @memberof Components/Helpers
 */
export function replaceBlock({ region, block, Component }) {
  const previousBlock = BlocksTable[region][block];

  if (!previousBlock) {
    throw new Error(`Block '${block}' of region ${region} not found. Use registerComponent to create it.`);
  }

  return registerBlock({
    name: block,
    region,
    Component
  });
}

/**
 * @name getBlockComponent
 * @method
 * @summary Get the Component registered within a block
 * @param {String} regionName The name of the block region.
 * @param {String} blockName The name of the block component to get.
 * @returns {Function|React.Component} A React component
 * @memberof Components/Helpers
 */
export const getBlockComponent = (regionName, blockName) => BlocksTable[regionName][blockName].Component;

/**
 * @name loadRegisteredBlocks
 * @method
 * @summary Populate the final BlockComponents object with the contents of the lookup table.
 * This should only be called once on app startup.
 * @returns {Object} An object containing all of the registered blocks by region
 * @memberof Components/Helpers
 **/
export function loadRegisteredBlocks() {
  Object.keys(BlocksTable).forEach((regionName) => {
    BlockComponents[regionName] = getBlocks(regionName);
  });
  return BlockComponents;
}
