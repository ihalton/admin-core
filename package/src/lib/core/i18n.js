import i18n from "i18next";
import { initReactI18next } from "react-i18next";

/**
 * Init i18next
 * @returns {undefined}
 */
export function initI18next() {
  i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
      backend: {
        loadPath: "/public/locales/{{lng}}-{{ns}}.json"
      },
      resources: {
        en: {
          translation: {
            shop: "Shop"
          }
        }
      },
      lng: "en",
      fallbackLng: "en",

      interpolation: {
        escapeValue: false
      }
    });
}
