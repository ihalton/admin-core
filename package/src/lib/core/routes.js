export const routes = [];
export const defaultRouteGroups = {
  navigation: "navigation",
  settings: "settings"
};

/**
 * @name registerRoute
 * @summary Registers new route for the admin UI.
 * @param {Object} route - The route
 * @param {String} route.path - The URL path for this route
 * @param {Node|String} route.MainComponent - A react component to render in
 * the main content area or the name of a Blaze template that has been registered
 * by a package.
 * @param {Node} route.IconComponent - A React component that renders the menu icon for this route
 * @param {String} route.navigationItemLabel - The i18n key for this route, i.e. "admin.dashboard.ordersLabel"
 * @returns {undefined}
 */
export function registerRoute(route) {
  routes.push(route);
}
