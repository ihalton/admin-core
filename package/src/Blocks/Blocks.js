import React from "react";
import PropTypes from "prop-types";
import { BlockComponents } from "../lib/core/blocks";

/**
 * @name Blocks
 * @method
 * @memberof Components/Helpers
 * @param {Object} props Component props
 * @returns {React.Component} returns a React component containing requested blocks
 */
export function Blocks(props) {
  const {
    region,
    children,
    blockProps
  } = props;
  const blocks = BlockComponents[region];
  if (!blocks) return null;

  const elements = blocks.map((BlockComponent, key) => (
    <BlockComponent key={key} {...blockProps} />
  ));

  return children(elements);
}

Blocks.defaultProps = {
  children: (blocks) => blocks
};

Blocks.propTypes = {
  children: PropTypes.func.isRequired
};

export default Blocks;
