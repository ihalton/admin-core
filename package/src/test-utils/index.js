
import React from "react";
import PropTypes from "prop-types";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@material-ui/core";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider } from "@apollo/react-testing";
import { defaultTheme } from "@reactioncommerce/catalyst";
import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import apolloMocks from "./apolloMocks";

i18n
  .use(initReactI18next)
  .init({
    resources: {
      en: {
        translation: {}
      }
    },
    lng: "en",
    fallbackLng: "en",

    interpolation: {
      escapeValue: false
    }
  });

/**
 * Component that wraps components with mock providers during testing.
 * @return {Component} - Component wrapped with mock providers
 */
const TestProviders = ({ children }) => (
  <MockedProvider mocks={apolloMocks}>
    <ThemeProvider theme={defaultTheme}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </ThemeProvider>
  </MockedProvider>
);

TestProviders.propTypes = {
  /** React Component */
  children: PropTypes.element.isRequired
};

/**
 * Custom test renderer that wraps all components with the appropriate mock providers.
 * @param {Component} component - React component to render.
 * @param {Object} options - Options.
 * @return {Object} - @see {@link https://testing-library.com/docs/react-testing-library/api#render-result|react-testing-library}
 */
const renderWithProviders = (component, options) => (
  render(component, { wrapper: TestProviders, ...options })
);

export * from "@testing-library/react";
export { renderWithProviders as render };
