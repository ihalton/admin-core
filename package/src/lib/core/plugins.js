import { registerRoute } from "./routes";
import { registerBlock, replaceBlock } from "./blocks";

export const plugins = [];

/**
 * Plugin registration
 * @param {Plugin} plugin An object containing the plugins configuration
 * @returns {undefined} no return
 */
export function registerPlugin(plugin) {
  plugins.push(plugin({
    registerRoute,
    registerBlock,
    replaceBlock
  }));
}
