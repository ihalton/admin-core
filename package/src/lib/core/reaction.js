import React from "react";
import ReactDOM from "react-dom";
import SettingsIcon from "mdi-material-ui/Settings";
import initApollo from "../graphql/initApollo";
import SettingsDashboard from "../../SettingsDashboard";
import Profile from "../../Profile";
import { getOidcProps } from "./authentication";
import { plugins } from "./plugins";
import { registerRoute } from "./routes";
import { initI18next } from "./i18n";
import { loadRegisteredBlocks } from "./blocks";

/**
 * The starting point for the web-app
 * @param {Object} props Props and config for the app
 * @returns {undefined} no return
 */
export async function Reaction(props) {
  const {
    config: {
      PUBLIC_GRAPHQL_API_URL,
      PUBLIC_I18N_BASE_URL,
      PUBLIC_I18N_FALLBACK_LANGUAGE,
      PUBLIC_OIDC_CLIENT_ID,
      PUBLIC_OIDC_URL,
      PUBLIC_ROOT_URL
    },
    AppComponent,
    DashboardComponent,
    SettingsDashboardComponent = SettingsDashboard,
    dashboardComponentProps,
    settingsRouteProps,
    shouldShowSettingsInNavigation = true
  } = props;

  // Initialize apollo client to be used for the ApolloProvider in the AppComponent
  const apolloClient = initApollo({ graphqlApiUrl: PUBLIC_GRAPHQL_API_URL });

  // Init i18next
  await initI18next({
    i18nBaseUrl: PUBLIC_I18N_BASE_URL,
    fallbackLng: PUBLIC_I18N_FALLBACK_LANGUAGE
  });

  // Load registered blocks
  loadRegisteredBlocks();

  if (shouldShowSettingsInNavigation) {
    registerRoute({
      group: "navigation",
      priority: 9999,
      path: "/settings/:setting?",
      href: "/settings/shop",
      LayoutComponent: null,
      MainComponent: SettingsDashboardComponent,
      IconComponent: SettingsIcon,
      navigationItemLabel: "admin.settings.settingsLabel",
      ...settingsRouteProps
    });
  }

  // Register a profile page
  registerRoute({
    path: "/profile",
    MainComponent: Profile
  });

  // Create OIDC props to be used on the AuthenticationProvider in the AppComponent
  const authenticationProviderProps = getOidcProps({
    oidcClientId: PUBLIC_OIDC_CLIENT_ID,
    oidcUrl: PUBLIC_OIDC_URL,
    rootUrl: PUBLIC_ROOT_URL
  });

  ReactDOM.render(
    (
      <AppComponent
        DashboardComponent={DashboardComponent}
        apolloClient={apolloClient}
        authenticationProviderProps={authenticationProviderProps}
        dashboardComponentProps={dashboardComponentProps}
        plugins={plugins}
      />
    ), document.getElementById("app")
  );
}
