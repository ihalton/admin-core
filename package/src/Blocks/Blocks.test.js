/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
/* eslint-disable react/no-multi-comp */
import React from "react";
import { render } from "../test-utils";
import { registerBlock } from "../lib/core/blocks";
import Blocks from "./Blocks";

registerBlock({
  name: "block-1",
  region: "test",
  Component: () => (<span>Block 1</span>)
});

registerBlock({
  name: "block-2",
  region: "test",
  Component: () => (<span>Block 2</span>)
});

test("render the Blocks region named 'test' with two registered blocks", () => {
  const { asFragment } = render((
    <Blocks region="test" />
  ));
  expect(asFragment()).toMatchSnapshot();
});
