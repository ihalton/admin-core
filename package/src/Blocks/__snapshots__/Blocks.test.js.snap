// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the Blocks region named 'test' with two registered blocks 1`] = `<DocumentFragment />`;
